File: java/src/datastructure/stack/LinkedStack.java
package datastructure.stack;

import java.util.EmptyStackException;

/**
 * @author roseduan[roseduan520@gmail.com]
 * @time 2019/10/27 15:28
 * @description 使用链表实现栈
 */
public class LinkedStack {

    private ListNode head;
    private int size;

    public LinkedStack(int val) {
        this.head = new ListNode(val);
        this.size = 1;
    }

    private static class ListNode{
        int val;
        ListNode next;

        ListNode(int val) {
            this.val = val;
            this.next = null;
        }
    }

    /**
     * 入栈
     */
    public void push(int value) {
        ListNode node = new ListNode(value);
        size++;

        if (head == null) {
            head = node;
        }
        else {
            node.next = head;
            head = node;
        }
    }

    /**
     * 出栈
     */
    public int pop(){
        if (head == null){
            throw new EmptyStackException();
        }

        ListNode result = head;
        head = head.next;
        size--;

        return result.val;
    }

    /**
     * 获取栈顶元素
     */
    public int peek(){
        if (head == null){
            throw new EmptyStackException();
        }

        return head.val;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        ListNode p = head;

        while (p != null){
            sb.append(p.val).append(" ");
            p = p.next;
        }

        return sb.toString();
    }
}

File: go/stack/linked_stack/linked_stack.go
package linked_stack

import (
	"errors"
	"fmt"
)

//链式栈的实现(基于链表)

type ListNode struct {
	next *ListNode
	val  interface{}
}

type LinkedStack struct {
	head *ListNode
	size int
}

func NewLinkedStack() *LinkedStack {
	return &LinkedStack{
		head: nil,
		size: 0,
	}
}

//压栈
func (stack *LinkedStack) Push(val interface{}) {
	node := &ListNode{nil, val}
	if stack.Empty() {
		stack.head = node
	} else {
		node.next = stack.head
		stack.head = node
	}

	stack.size++
}

//出栈
func (stack *LinkedStack) Pop() (interface{}, error) {
	if stack.Empty() {
		return nil, errors.New("stack is empty")
	}

	result := stack.head.val
	stack.head = stack.head.next
	stack.size--

	return result, nil
}

//获取栈顶元素
func (stack *LinkedStack) Peek() (interface{}, error) {
	if stack.Empty() {
		return nil, errors.New("stack is empty")
	}
	return stack.head.val, nil
}

func (stack *LinkedStack) Empty() bool {
	return stack.size <= 0
}

func (stack *LinkedStack) Size() int {
	return stack.size
}

func (stack *LinkedStack) PrintData() {
	if stack.Empty() {
		fmt.Println("stack is empty")
	} else {
		for p := stack.head; p != nil; p = p.next {
			fmt.Print(p.val, " ")
		}
		fmt.Println()
	}
}
